### ISSUE TITLE
datastore: Struct tag to Omit Empty

### ISSUE THREAD

**Issue by patrickml** [GoogleCloudPlatform/google-cloud-go#418]
So there doesn't seem to be a way to omit keys from a struct that are being saved to Datastore if the values are empty. The `json.Unmarshal` has the tag "omitempty" I am wondering if there is a way we can add this to the SDK for `datastore:"keyName,omitempty"`. 

We have a use case where we will have millions of records that may or may not have a certain key on them. I'd hate to see there be all of these empty keys be indexed if they do not exist. Maybe I am going about this wrong, but I could see this being a valuable concept to save storage and memory.

Example
```go
type MyStruct struct {
	SomeID string                 `datastore:"someId,omitempty" json:"someId,omitempty"`
}
```

Current Solution:

I believe a current solution is the use the PropertyLoadSaver to omit the keys, but in the long run it could be much simpler to just do through tags.

**Comment by lukatendai** [comment]
very useful, when it will be added?

**Comment by adams-sarah** [comment]
@patrickml just to clarify before I start implementation, you are hoping to use the `omitempty` tag option to completely omit empty struct fields from the Entity on creation. Not to partially-update an Entity, as is often the use-case for `omitempty` with the JSON package for an HTTP-based REST API.

Is that correct?

**Comment by lukatendai** [comment]
@adams-sarah sounds right to me... this is the comment from JSON package:

`
// The "omitempty" option specifies that the field should be omitted
// from the encoding if the field has an empty value, defined as
// false, 0, a nil pointer, a nil interface value, and any empty array,
// slice, map, or string.
`

**Comment by patrickml** [comment]
@adams-sarah correct for example

If given this struct
```go
type MyStruct struct {
	ID string                            `datastore:"id" json:"id"`,
        NickName string             `datastore:"nickName,omitempty" json:"nickName,omitempty"`,
        Name string                    `datastore:"name" json:"name"`,
}
```

When I initialize it without NickName
```go
m := MyStruct{
   Name: "John"
}
```

Then I should expect that `ID` and `Name` are stored with a key, but `NickName` should not exist as a key on the document

```js
{
   id: "",
   name: "John"
}
```

**Comment by adams-sarah** [comment]
Excellent. Thanks for clarifying. I will start on this today.

**Comment by okdave** [comment]
Sounds like a good feature – you'll just need to be really clear what effect it has on sub-structs. I'm guessing something like `noindex` would work well there?

**Comment by adams-sarah** [comment]
@okdave Well, that would definitely make life simpler. But I am not sure the same behavior of propagating the tag down to substructs makes sense logically. That would mean that if you had an `omitempty` tag on one parent struct field, all fields of all substructs and their substructs would be omitted if empty. Do you think that's the right behavior?

**Comment by okdave** [comment]
Yeah, I'm not sure. Seems like the alternatives are:

1. all sub-fields are omitted if empty – easy to implement, consistent with `noindex`
2. the sub-struct is only omitted if all its fields are zero – seems difficult to implement, and tough to understand (maybe easier with pointers-to-structs though)?
3. `omitempty` is not supported at all on sub-structs – sounds good, but I'd fail hard so you can change your mind later

Either (1) or (3) seem sensible. I can't think of any other possible behaviours, but I'm sure there are other options.

**Comment by jba** [comment]
tl;dr: (3)

I now think that the right question to ask when dealing with struct mapping issues like this is: "What would json do?" encoding/json, that is. We should strive to match its behavior, since it's a heavily used package in the standard library. 

https://play.golang.org/p/1XF9MP8wC0 shows encoding/json's "omitempty" behavior on:
* a field of type slice of structs
* a field of type struct 
* an embedded struct without a name tag
* an embedded struct with a name tag

(The last two are different because encoding/json treats an embedded struct with a name tag just like an ordinary named field of struct type.)

omitempty doesn't apply to a one-element slice even if that one element is an empty struct. It also doesn't apply in the other three cases.

This behavior is consistent with the documentation of Unmarshal, which says:

>Each exported struct field becomes a member of the object unless...the field is empty and its tag specifies the "omitempty" option. ... The empty values are false, 0, any nil pointer or interface value, and any array, slice, map, or string of length zero. 

It doesn't talk about structs in that list, so we can assume they're never "empty". (I wish the doc was more explicit about that, though.)

The implementation is in the [isEmptyValue function](https://golang.org/src/encoding/json/encode.go#L288) if you're curious.

**Comment by jba** [comment]
By the way, I agree with @okdave that we should return an error if omitempty is used for a type that can never be empty, rather than just ignoring it.

**Comment by adams-sarah** [comment]
@jba returning an error if the type could never be empty is not the behavior of encoding/json.
I agree that we should keep the behavior of the two identical, and thus I do not think we should return an error. 
I have documented the types which can be empty in doc.go, and that the behavior is identical to encoding/json.
